lang/src/org/partiql/lang/eval/physical/operators/AggregationFunctionFactory.kt

```diff
@@ -0,0 +1,203 @@
+import org.partiql.lang.eval.SequenceExprValue
+import org.partiql.lang.eval.errNoContext
+import org.partiql.lang.eval.isUnknown
+import org.partiql.lang.eval.numberValue
+import org.partiql.lang.eval.physical.EvaluatorState
+import org.partiql.lang.eval.physical.SetVariableFunc
+import org.partiql.lang.util.bigDecimalOf
+import org.partiql.lang.util.div
+import org.partiql.lang.util.plus
+import org.partiql.lang.util.toIntExact
+import java.math.BigDecimal
+
+public interface AggregationFunctionFactory {
+    public val key: AggregationFunctionFactoryKey
+
+    fun create(
+        impl: PartiqlPhysical.Impl,
+        variables: Array<PartiqlPhysical.VarDecl>
+    ): SetVariableFunc
+}
+
+internal class SumFactory(name: String) : AggregationFunctionFactory {
+    override val key: AggregationFunctionFactoryKey = AggregationFunctionFactoryKey(AggregationKind.SUM, name)
+    override fun create(
+        impl: PartiqlPhysical.Impl,
+        variables: Array<PartiqlPhysical.VarDecl>
+    ): SetVariableFunc {
+        val index = variables[0].index.value.toIntExact()
+        return thunk@{ state, value ->
+            if (value.isUnknown()) {
+                initializeMissingRegisterWithNull(state, index)
+                return@thunk
+            }
+            checkIsNumberType(funcName = "SUM", value = value)
+            conditionallyInitializeRegister(state, index, state.valueFactory.newInt(0))
+            state.registers[index] = (state.registers[index].numberValue() + value.numberValue()).exprValue(state.valueFactory)
+        }
+    }
+}
+
+internal class AverageFactory(name: String) : AggregationFunctionFactory {
+    override val key: AggregationFunctionFactoryKey = AggregationFunctionFactoryKey(AggregationKind.SUM, name)
+    override fun create(
+        impl: PartiqlPhysical.Impl,
+        variables: Array<PartiqlPhysical.VarDecl>
+    ): SetVariableFunc {
+        val (indexAvg, indexSum, indexCount) = variables.map { varDecl -> varDecl.index.value.toIntExact() }
+        return thunk@{ state, value ->
+            if (value.isUnknown()) {
+                initializeMissingRegisterWithNull(state, indexAvg)
+                return@thunk
+            }
+            checkIsNumberType(funcName = "AVG", value = value)
+            conditionallyInitializeRegister(state, indexAvg, state.valueFactory.newInt(0))
+            conditionallyInitializeRegister(state, indexSum, state.valueFactory.newInt(0))
+            conditionallyInitializeRegister(state, indexCount, state.valueFactory.newInt(0))
+            state.registers[indexSum] = (state.registers[indexSum].numberValue() + value.numberValue()).exprValue(state.valueFactory)
+            state.registers[indexCount] = (state.registers[indexCount].numberValue() + 1L).exprValue(state.valueFactory)
+            state.registers[indexAvg] = (state.registers[indexSum].numberValue() / bigDecimalOf(state.registers[indexCount].numberValue())).exprValue(state.valueFactory)
```

Comment:
A little bit counter-intuitive as we store intermediate result into the registers, and even more so with `AVG`...
As a by-product, this is a lot of local variables created. 
That said, of the top of my head, I can't really think of another way to do so without a large refactor... 
May be worth to add some comments explains this approach?

Reply:
Yeah, this is related to Partial vs Final vs Full Aggregation. I'll leave a comment.